### DISCUSSION TITLE
Reorganizing the project settings

### DISCUSSION THREAD

**Discussion by mihe** [godot-jolt/godot-jolt#1022] [General]
@jrouwe, I will likely be making a pull request to Godot next week for migrating this extension to a Godot engine module, and while doing so I thought it might be a good opportunity to look over the "public interface" of it and see if there are any things that ideally should be changed while we have the chance, and I wanted your input on it.

The initial merge of this integration will omit the Jolt-specific joint nodes, so really the only additions to Godot's physics interface will be the Jolt project settings, which will likely continue to live on under the "Jolt 3D" settings category[^1]. The joint interfaces will instead be addressed at a later point, by allowing the physics servers themselves to add/remove/modify properties (and maybe methods later on) for Godot's existing physics nodes/resources instead.

As you've seen already I took some creative freedoms when naming/grouping some of the project settings in this extension, and frankly I think I ended up regretting almost every instance where I deviated from what you called them. 😅

It started out mostly as me trying to "Godot-ify" the name of your settings, so (for example) instead of calling it "Penetration Slop" I named it "Contact Allowed Penetration" to line up with Godot Physics, which refers to it as "Contact Max Allowed Penetration". Then there were ones like the Baumgarte stabilization factor, which I for some reason felt needed to be configurable as a percentage rather than a factor, and having a stabilization percentage that was effectively less stable at 100% seemed odd, so I named it "Position Correction" instead.

The name of the groups ("Collisions", "Solver", etc.) are also up for changing. These also came out of wanting to mimick the Godot Physics project settings, which has a "Solver" category, but I'm obviously not that familiar with the inner workings of Jolt (or physics engines in general) so knowing what belongs in a "Solver" category and what doesn't has been tricky.

I'm not sure these "reimaginings" really helped anyone, and instead probably just made it confusing for anyone who wanted to look through your code/documentation/issues/PRs to learn more about Jolt.

Anyway, so we currently have these project settings:

<details><summary><b>Click to expand/collapse</b></summary>
<p>

- `sleep/enabled`
  - Maps to `JPH::PhysicsSettings::mAllowSleeping`.
- `sleep/velocity_threshold`
  - Maps to `JPH::PhysicsSettings::mPointVelocitySleepThreshold`.
- `sleep/time_threshold`
  - Maps to `JPH::PhysicsSettings::mTimeBeforeSleep`.
- `collisions/use_shape_margins`
  - Sets `mConvexRadius` to 0 for all applicable shapes when disabled.
- `collisions/use_enhanced_internal_edge_removal`
  - Maps to `JPH::BodyCreationSettings::mEnhancedInternalEdgeRemoval`.
- `collisions/areas_detect_static_bodies`
  - Maps to `JPH::BodyCreationSettings::mCollideKinematicVsNonDynamic`.
- `collisions/report_all_kinematic_contacts`
  - Maps to `JPH::BodyCreationSettings::mCollideKinematicVsNonDynamic`.
- `collisions/soft_body_point_margin`
  - Maps to `JPH::SoftBodySharedSettings::mVertexRadius`.
- `collisions/body_pair_cache_enabled`
  - Maps to `JPH::PhysicsSettings::mUseBodyPairContactCache`.
- `collisions/body_pair_cache_distance_threshold`
  - Maps to `JPH::PhysicsSettings::mBodyPairCacheMaxDeltaPositionSq`.
- `collisions/body_pair_cache_angle_threshold`
  - Maps to `JPH::PhysicsSettings::mBodyPairCacheCosMaxDeltaRotationDiv2`.
- `joints/world_node`
  - Controls the ordering of `JPH::Body::sFixedToWorld` for single-body constraints.
- `continuous_cd/movement_threshold`
  - Maps to `JPH::PhysicsSettings::mLinearCastThreshold`.
- `continuous_cd/max_penetration`
  - Maps to `JPH::PhysicsSettings::mLinearCastMaxPenetration`.
- `kinematics/use_enhanced_internal_edge_removal`
  - Controls the use of `JPH::NarrowPhaseQuery::CollideShapeWithInternalEdgeRemoval`.
- `kinematics/recovery_iterations`
  - Used to control the behavior of `move_and_collide`.
- `kinematics/recovery_amount`
  - Used to control the behavior of `move_and_collide`.
- `queries/use_enhanced_internal_edge_removal`
  - Controls the use of `JPH::NarrowPhaseQuery::CollideShapeWithInternalEdgeRemoval`.
- `queries/enable_ray_cast_face_index`
  - Maps to `JPH::MeshShapeSettings::mPerTriangleUserData`.
- `solver/position_iterations`
  - Maps to `JPH::PhysicsSettings::mNumPositionSteps`.
- `solver/velocity_iterations`
  - Maps to `JPH::PhysicsSettings::mNumVelocitySteps`.
- `solver/position_correction`
  - Maps to `JPH::PhysicsSettings::mBaumgarte`.
- `solver/active_edge_threshold`
  - Maps to `JPH::MeshShapeSettings::mActiveEdgeCosThresholdAngle` (and height field).
- `solver/bounce_velocity_threshold`
  - Maps to `JPH::PhysicsSettings::mMinVelocityForRestitution`.
- `solver/contact_speculative_distance`
  - Maps to `JPH::PhysicsSettings::mSpeculativeContactDistance`.
- `solver/contact_allowed_penetration`
  - Maps to `JPH::PhysicsSettings::mPenetrationSlop`.
- `limits/world_boundary_shape_size`
  - Maps to `JPH::PlaneShapeSettings::mHalfExtent`.
- `limits/max_linear_velocity`
  - Maps to `JPH::BodyCreationSettings::mMaxLinearVelocity`.
- `limits/max_angular_velocity`
  - Maps to `JPH::BodyCreationSettings::mMaxAngularVelocity`.
- `limits/max_bodies`
  - Maps to the parameter in `JPH::PhysicsSystem::Init`.
- `limits/max_body_pairs`
  - Maps to the parameter in `JPH::PhysicsSystem::Init`.
- `limits/max_contact_constraints`
  - Maps to the parameter in `JPH::PhysicsSystem::Init`.
- `limits/max_temporary_memory`
  - Sets the allocation size of `JoltTempAllocator`.

</p>
</details> 

I had a few potential changes in mind already, like perhaps:

1. Rename the `collisions/...` category to `bodies/...`.
2. Move the `sleep/...` and `continuous_cd/...` settings into `bodies/...`.
3. Rename `soft_body_point_margin` to `soft_body_point_radius`.
4. Change `position_correction` to be `baumgarte_stabilization_factor` and have it not be a percentage.
5. Change `solver/contact_allowed_penetration` to `bodies/penetration_slop`.
6. Change `contact_speculative_distance` to `speculative_contact_distance`.
7. Rename `max_temporary_memory` to `temporary_memory_buffer_size`.

None of these are written in stone of course. I'm curious if/how you would name/group the settings any differently.

I'm also curious what you think of keeping `use_shape_margins` around. It's been brought up several times as a source of potential problems, but at the same time it seems to have become a relatively popular option to disable as well, as a workaround for Godot's lack of actual surface normals in its query results.

It should also be noted that once the physics servers in Godot gain the ability to add properties to the physics nodes/resources then a couple of these settings will be removed in favor of per-shape/per-body properties instead. These would include `use_enhanced_internal_edge_removal`, `areas_detect_static_bodies`, `report_all_kinematic_contacts`, `soft_body_point_margin`, `enable_ray_cast_face_index`, `active_edge_threshold` and `world_boundary_shape_size`.

[^1]: There's currently a [pending](http://github.com/godotengine/godot/pull/98528) pull request which will separate out the Godot Physics-specific project settings from the `physics/3d/...` category, leaving only the stuff that's more or less certain to be applicable to all physics engines.

**Comment by mihe** [comment]
> I'm also curious what you think of keeping `use_shape_margins` around. It's been brought up several times as a source of potential problems, but at the same time it seems to have become a relatively popular option to disable as well, as a workaround for Godot's lack of actual surface normals in its query results.

I was thinking maybe this could be changed to not be a boolean, but instead expose the fraction by which the extent of the shape is multiplied (i.e. what was suggested [here](https://github.com/godot-jolt/godot-jolt/issues/379#issuecomment-1566133839) and is currently [hardcoded](https://github.com/godot-jolt/godot-jolt/blob/8f1212e19a1bdc76ce34b5ddcfabbfd6e409ef9c/src/shapes/jolt_box_shape_impl_3d.cpp#L7) in various places) so you'd have something like a `shape_margin_fraction` project setting instead, which could then be lowered to a smaller value instead of being set to 0, to hopefully mitigate some of the problems that arise from setting it to 0.

This is also how Unreal does it with Chaos, which exposes a setting called "Collision Margin Fraction" that serves the exact same purpose, as seen [here](https://dev.epicgames.com/documentation/en-us/unreal-engine/physics-settings-in-the-unreal-engine-project-settings).

**Reply by jrouwe** [reply]
Using a margin factor probably makes sense for:

```C++
JPH::ShapeRefC JoltBoxShapeImpl3D::_build() const {
	const auto min_half_extent = (float)half_extents[half_extents.min_axis_index()];
	const float actual_margin = MIN(margin, min_half_extent * JoltProjectSettings::get_margin_factor());
        ...
```

if you set the default to the 0.08 of `MARGIN_FACTOR`. But then you'd need to make it consistent for the other shapes as well:

```C++
JPH::ShapeRefC JoltConvexPolygonShapeImpl3D::_build() const {
        ...
	const auto min_half_extent = (float)half_extents[half_extents.min_axis_index()];
	const float actual_margin = MIN(margin, min_half_extent * JoltProjectSettings::get_margin_factor());
```

but since the bounding box of a convex polygon is only a coarse approximation, the parameter is much less predictable/useful there.

I think it's a better setting than `use_shape_margins`, but nothing beats setting the margin per shape (which is also still possible I guess).

**Reply by mihe** [reply]
Good catch. I had glossed over how this would look for convex hulls.

> I think it's a better setting than `use_shape_margins`

Great, I'll go ahead and make the change then.

> nothing beats setting the margin per shape (which is also still possible I guess)

Yes.

**Comment by jrouwe** [comment]
Naming things is one of the hardest things to do. To be honest, I don't think the names you have are bad at the moment. The only thing that I find confusing is the category `kinematics` as it seems to only apply to a very specific subset of functions, I would propose something like `character`.

> Rename the collisions/... category to bodies/....

I'm not sure if that is a better name. I would suggest `simulation` or `simulation/collisions` instead (to distinguish between collision queries and collisions during simulation). `use_shape_margins` is the only setting that also affects collision queries, so maybe it doesn't belong there.

> Move the sleep/... and continuous_cd/... settings into bodies/....

Perhaps as sub categories of `simulation`?

> Rename soft_body_point_margin to soft_body_point_radius.

Ok

> Change position_correction to be baumgarte_stabilization_factor and have it not be a percentage.

Ok

> Change solver/contact_allowed_penetration to bodies/penetration_slop.

Maybe `simulation/penetration_slop`. Perhaps `solver` should be merged into `simulation` as well? Or everything in `solver` could be a sub category: `simulation/solver`.

> Change contact_speculative_distance to speculative_contact_distance.

Ok

> Rename max_temporary_memory to temporary_memory_buffer_size.

Ok

**Reply by mihe** [reply]
> Naming things is one of the hardest things to do.

Most definitely!

> The only thing that I find confusing is the category `kinematics` as it seems to only apply to a very specific subset of functions, I would propose something like `character`.

Unfortunately these methods are applicable to more than just `CharacterBody3D`, although that's certainly where they're most relevant, through its [`move_and_slide`](https://docs.godotengine.org/en/latest/classes/class_characterbody3d.html#class-characterbody3d-method-move-and-slide) method, which is indirectly powered by the [`body_test_motion`](https://docs.godotengine.org/en/latest/classes/class_physicsserver3d.html#class-physicsserver3d-method-body-test-motion) method that I implement from the extension.

[`move_and_collide`](https://docs.godotengine.org/en/latest/classes/class_physicsbody3d.html#class-physicsbody3d-method-move-and-collide) for example, which is a thin wrapper around `body_test_motion`, is a method on the `PhysicsBody3D` node, which is the base class for all physics bodies (except `SoftBody3D`), so you can in theory "freeze" a `RigidBody3D` as kinematic and move it around in a similar fashion with `move_and_collide`, where these `kinematics/...` settings would then also apply. The same would also apply to the already kinematic `AnimatableBody3D`.

If the name `kinematics` still seems off then I'm open to other suggestions. Maybe something like `kinematic_movement`, to be more specific, or would that sound redundant?

> I'm not sure if that is a better name. I would suggest `simulation` or `simulation/collisions` instead (to distinguish between collision queries and collisions during simulation). `use_shape_margins` is the only setting that also affects collision queries, so maybe it doesn't belong there.

`simulation` does indeed sound better than `bodies`.

I'll experiment with the sub-categories. As far as I'm aware there's no project setting in Godot currently that goes beyond three levels of nesting, so `physics/jolt_3d/some_category/...` would be the "limit". It does technically seem to work to add another, but it looks a little bit off.

> Perhaps as sub categories of `simulation`?

For the reason mentioned above, these could probably also have their sub-category in the name itself, so you'd have `simulation/allow_sleep`, `simulation/sleep_point_velocity_threshold`, `simulation/ccd_movement_threshold`, `simulation/ccd_max_penetration`, etc.

> Maybe `simulation/penetration_slop`. Perhaps `solver` should be merged into `simulation` as well? Or everything in `solver` could be a sub category: `simulation/solver`.

I'm somewhat keen on getting rid of the `solver` category, so merging it into `simulation` sounds like a good idea, along with `penetration_slop`.

> `use_shape_margins` is the only setting that also affects collision queries, so maybe it doesn't belong there.

I imagine this also applies to `simulation/active_edge_threshold`. Perhaps these two could live under the `collisions/...` category, to hopefully communicate that they apply to both simulation and query collisions? Although, as mentioned above, it might be that `active_edge_threshold` is removed eventually in favor of a per-shape property.

**Reply by jrouwe** [reply]
> If the name kinematics still seems off then I'm open to other suggestions. Maybe something like kinematic_movement, to be more specific, or would that sound redundant?

To me the word 'kinematic' means 'an object that moves during the simulation step with constant velocity'. Anything that is updated by the user outside of the simulation step or an object that is affected in its velocity by colliding is not really a kinematic object.

**Reply by mihe** [reply]
> To me the word 'kinematic' means 'an object that moves during the simulation step with constant velocity'. Anything that is updated by the user outside of the simulation step or an object that is affected in its velocity by colliding is not really a kinematic object.

That's fair. There is however some precedence for referring to this as "kinematic" in Godot already, which has named the result struct of `move_and_collide` as `KinematicCollision3D`, but I suppose that also might not strictly fit the physics definition.

Perhaps `motion_queries` is a better name for this category. It might not be immediately obvious that the settings would then apply to `move_and_slide`, since `move_and_slide` isn't really a query in the same sense and instead just performs the movement right away, but `kinematics` was never really self-explanatory either I guess. It can be made clear in the tooltips of the respective settings what methods they apply to.

EDIT: Also, just to clarify, while these methods can technically be used by non-kinematic objects, I imagine that was never the intention. Your point about it happening outside of the simulation still applies though.

**Comment by qbby** [comment]
Does this mean the "legacy ray cast" project setting will be removed in the transition to a Godot module? That behavior was the logical solution for certain use cases, so I'd be sad to see it go.

**Reply by mihe** [reply]
Yes, it's been removed from the module (see pull request [here](https://github.com/godotengine/godot/pull/99895)) and the behavior of Godot Physics is now the only one. I agree that the "legacy" behavior made more sense (hence the divergence initially) but it wasn't anything requested by the community to begin with, so I figured if people really want it then it should go the route of becoming a Godot proposal, and be changed in both engines.

**Reply by qbby** [reply]
Well, as it stands, I'm not sure what the Godot Physics version of hit_back_faces is even useful for, and I'm not optimistic that it would be addressed anytime soon. A proposal exists, but hasn't received any attention for 4 months. It seems nontrivial to implement that behavior in Godot Physics, while it was natural in Jolt.
In the meantime, if the Godot maintainers didn't request the removal of that setting, then I don't see the problem with allowing the user to choose. The recent separation of Godot Physics project settings implies that physics engines are expected to implement different settings, anyway.
Also, I don't think the community realizes the use cases of this feature, so it would be overlooked. For example, it can be a foolproof way to stop a movement at the exact point when a character would leave the ground. I'm not sure how else you would cleanly and generically accomplish this.
Either way, this project is great work. Kudos!
